source/agora/script/Opcodes.d

```diff
@@ -0,0 +1,167 @@
+/// without having to encode it in a separate length byte. 64 was chosen as
+/// the upper bound because our signatures are 64 bytes.
+/// For pushes of data longer than 64 bytes use the `PUSH_DATA_*` opcodes.
+enum OP : ubyte
+{
+    /// Using this is an error and will invalidate the transaction
+    /// Purposefully located first to default OPs to errors.
+    INVALID = 0x45,
+
+    /// Used to encode small length of data to push to the stack (up to 64 bytes),
+    /// may be used with `case PUSH_BYTES_1: .. case PUSH_BYTES_64:` syntax.
+    PUSH_BYTES_1 = 0x01,
+    PUSH_BYTES_64 = 0x40, // 64 decimal
+
+    /// The next 1 byte contains the number of bytes to push onto the stack
+    PUSH_DATA_1 = 0x41,
+
+    /// The next 2 bytes (ushort in LE format) contains the number of bytes to
+    /// push onto the stack
+    PUSH_DATA_2 = 0x42,
+
+    /// The next 4 bytes (ushort in LE format) contains the number of bytes to
+    /// push onto the stack
+    RESERVED_PUSH_DATA_3 = 0x43,
+
+    /// Pushes True onto the stack
+    TRUE = 0x44,
+
+    /// Pushes False onto the stack
+    FALSE = 0x00,
+
+    /// Conditionals
+    IF = 0x46,
+    NOT_IF = 0x47,
+    ELSE = 0x48,
+    END_IF = 0x49,
+
+    /// Hash the value
+    HASH = 0x4A,
+
+    /// Duplicate the item on the stack
+    DUP = 0x4B,
+
+    /// Checks the stack item is equal to the input value
+    CHECK_EQUAL = 0x4C,
+
+    /// Verifies the stack item is equal to the input value,
+    /// fails the script if it's false
+    VERIFY_EQUAL = 0x4D,
+
+    /// Checks the signature with the given public key,
+    /// pushes the result to the stack
+    CHECK_SIG = 0x4E,
+
+    /// Checks the lock time against the current block height
+    VERIFY_LOCK_TIME = 0x4F,
+
+    /// Encodes a set of web assembly instructions
+    WEB_ASM = 0x50,
+}
+
+/*******************************************************************************
+
+    Converts the byte to an opcode,
+    or returns false if it's an unrecognized opcode.
+
+    Params:
+        value = the byte containing the opcode
+        opcode = will contain the opcode if it was recognized
+
+    Returns:
+        true if the value is a recognized opcode
+
+*******************************************************************************/
+
+public bool toOPCode (ubyte value, out OP opcode) pure nothrow @safe @nogc
+{
+    switch (value)
+    {
+        foreach (member; EnumMembers!OP)
+        {
+            case member:
+            {
+                opcode = member;
+                return true;
+            }
+        }
+
+        default:
+            break;
+    }
+
+    if (value >= 1 && value <= 64)  // PUSH_BYTES_1 .. PUSH_BYTES_64
+    {
+        opcode = cast(OP)value;  // dirty, but avoids need to define all pushes
+        return true;
+    }
+
+    return false;
+}
+
+/// Ditto, but assumes the opcode is valid (safe to use after validation)
+public OP toOPCode (ubyte value) pure nothrow @safe @nogc
```

Comment:
This is still problematic because client code could accidentally use `if (toOPCode(value)) { }`. Let's just get rid of it.